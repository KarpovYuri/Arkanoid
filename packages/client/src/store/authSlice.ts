import { createSlice } from '@reduxjs/toolkit';

const authSlice = createSlice({
  name: 'auth',
  initialState: {
    isLogged: false,
  },
  reducers: {
    setIsLogged(state, action) {
      state.isLogged = action.payload;
    },
  },
});

export const { setIsLogged } = authSlice.actions;
export default authSlice.reducer;
