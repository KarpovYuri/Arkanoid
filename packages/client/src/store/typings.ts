export type StateObject = {
  auth: {
    status: string;
    isLogged: boolean;
    error: string | null;
    userData: {
      id: number;
      first_name: string;
      second_name: string;
      display_name: string;
      login: string;
      avatar: string;
      email: string;
      phone: string;
    } | null;
    navLinks: {
      url: string,
      title: string,
      protect: string | boolean
    }[];
  },
};
