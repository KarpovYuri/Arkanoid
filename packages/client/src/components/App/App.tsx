import { StrictMode, useEffect } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Router } from '../../router/router';
import { Navigation } from '../Navigation';
import { ErrorBoundary } from '../ErrorBoundary';
import { useAppDispatch, useAppSelector } from '../../store/hooks';
import { getUserData } from '../../store/authSlice';
import { Loader } from '../Loader';

// useEffect(() => {
//   const fetchServerData = async () => {
//     const url = `http://localhost:${__SERVER_PORT__}`;
//     const response = await fetch(url);
//     const data = await response.json();
//     console.log(data);
//   };

//   fetchServerData();
// }, []);

export function App() {
  const dispatch = useAppDispatch();
  const { status } = useAppSelector((state) => state.auth);

  useEffect(() => {
    dispatch(getUserData());
  }, []);

  return (
    <StrictMode>
      <BrowserRouter>
        <ErrorBoundary>
          {status === 'loading'
            ? <Loader />
            : (
              <>
                <Navigation />
                <Router />
              </>
            )}
        </ErrorBoundary>
      </BrowserRouter>
    </StrictMode>
  );
}
