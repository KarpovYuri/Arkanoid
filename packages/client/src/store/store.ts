import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { authReducer } from './authSlice';

const rootReducer = combineReducers({
  // here we will be adding reducers
  // counter: counterReducer,
  auth: authReducer,
});

export const setupStore = configureStore({
  reducer: rootReducer,
});

// Use RootState and Dispatch typing inside files that will need this reference.
// Therefore, these definitions are created and exported within the store.ts file.

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof rootReducer>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof setupStore.dispatch;
