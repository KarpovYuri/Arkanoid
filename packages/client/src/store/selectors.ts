import { createSelector } from '@reduxjs/toolkit';
import { StateObject } from './typings';
import { NavLinkStatus } from '../utils/routeConstants';

export const selectUserData = (state: StateObject) => state.auth.userData;
export const selectAuthStatus = (state: StateObject) => state.auth.status;
export const selectIsLogged = (state: StateObject) => state.auth.isLogged;
export const selectAuthError = (state: StateObject) => state.auth.error;
export const selectNavlinks = (state: StateObject) => state.auth.navLinks;

export const selectNavLinksByIsLogged = createSelector(
  selectNavlinks,
  selectIsLogged,
  (allNavLinks, isLogged) => {
    if (isLogged) {
      return allNavLinks.filter((elem) => {
        if (elem.protect === NavLinkStatus.YES || elem.protect === NavLinkStatus.ALL) return elem;
        return false;
      });
    }
    return allNavLinks.filter((elem) => {
      if (elem.protect === NavLinkStatus.NO || elem.protect === NavLinkStatus.ALL) return elem;
      return false;
    });
  },
);
