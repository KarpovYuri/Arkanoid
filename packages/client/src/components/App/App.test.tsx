import { render, screen, waitForElementToBeRemoved } from '@testing-library/react';
// defineProperty for jest-test
// https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
import './matchMedia.mock'; // Must be imported before the tested file
import '@testing-library/jest-dom/extend-expect';

import { Provider } from 'react-redux';
import { AnyAction, configureStore, Store } from '@reduxjs/toolkit';

import { App } from './App';
import { rootReducer } from '../../store/store';

// @ts-ignore
global.fetch = jest.fn(() => Promise.resolve({ json: () => Promise.resolve('hey') }));

describe.skip('App component tests', () => {
  let store: Store<unknown, AnyAction>;
  beforeEach(() => {
    store = configureStore({ reducer: rootReducer });
  });

  test('Test the loader is shown', async () => {
    // TODO пока вставил спаны, чтобы не было ошибки
    // Как только закончу с SSR обещаю пересмотреть тест этого компонента)) (https://linear.app/22-ya-praktikum/issue/223-77/testy-na-komponent-app)
    render(<Provider store={store}><App><span>test</span></App></Provider>);

    const component = await screen.findByTestId('loader');
    expect(component).toBeInTheDocument();
    await waitForElementToBeRemoved(() => screen.getByTestId('loader'));
  });

  test('Test the navigation is shown', async () => {
    // TODO пока вставил спаны, чтобы не было ошибки
    // Как только закончу с SSR обещаю пересмотреть тест этого компонента)) (https://linear.app/22-ya-praktikum/issue/223-77/testy-na-komponent-app)
    render(<Provider store={store}><App><span>test</span></App></Provider>);
    const component = await screen.findByTestId('navigation');
    expect(component).toBeInTheDocument();
    expect(component).toMatchSnapshot();
  });
});
