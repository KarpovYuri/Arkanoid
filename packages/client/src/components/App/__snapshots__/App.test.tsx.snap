// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component tests Test the navigation is shown 1`] = `
<nav
  class="nav"
  data-testid="navigation"
>
  <div
    class="container"
  >
    <div
      class="navRow"
    >
      <button
        class="darkModeBtn"
        data-testid="toggleThemeBtn"
        type="button"
      >
        <img
          alt="Light mode"
          class="darkModeBtnIcon"
          src="default"
        />
        <img
          alt="Dark mode"
          class="darkModeBtnIcon"
          src="default"
        />
      </button>
      <ul
        class="navList"
      >
        <li
          class="navListItem"
        >
          <a
            aria-current="page"
            class="navListLink linkActive"
            href="/"
          >
            Главная
          </a>
        </li>
        <li
          class="navListItem"
        >
          <a
            class="navListLink"
            href="/authorization"
          >
            Вход
          </a>
        </li>
        <li
          class="navListItem"
        >
          <a
            class="navListLink"
            href="/registration"
          >
            Регистрация
          </a>
        </li>
      </ul>
    </div>
  </div>
</nav>
`;
